### ISSUE TITLE
Ghostscript 9.27+ removes .setfilladjust command used for "center of pixel" in gstoraster

### ISSUE THREAD

**Issue by kylemacfarlane** [OpenPrinting/cups-filters#164]
I'm not sure if you prefer bugs here or over on linuxfoundation...

A few years back https://github.com/OpenPrinting/cups-filters/commit/0090479c98a2200f326c056f641302d148bb93a4#diff-6e52c684ac817eafc9e3eb3a5cbe2d78 added a center of pixel command to improve the quality of barcodes from label printers. It depends on the undocumented `.setfilladjust` command in Ghostscript.

However it turns out that `.setfilladjust` was just a wrapper around `.setfilladjust2` and the wrapper has been removed in Ghostscript 9.27: http://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff;h=25c744f8a7761676d7790f513ca138eaaed2c366

So the solution going forwards is to change the command to `0 0 .setfilladjust2`.

But then we have another problem. It turns out that Ghostscript 9.22 removed access to `.setfilladjust2`: http://git.ghostscript.com/?p=ghostpdl.git;a=commitdiff;h=1497d65039885a52b598b137dd8622bd4672f9be

So in summary:

- `.setfilladjust` works in gs <= 9.26
- `.setfilladjust2` works in gs <= 9.21 and >= 9.27

So either you'd need to check the gs version or since center of pixel is a relatively niche feature that's not enabled by default maybe you could write off 9.22-9.26 and hope it doesn't cause too much disruption.

Good news is that at least in the 9.27 commit they said they intend to document `.setfilladjust2` so hopefully this won't happen again.

**Comment by tillkamppeter** [comment]
Fixed in commit e7103a7aaa. Thank you for the report.

**Comment by kylemacfarlane** [comment]
Thanks for fixing it so quickly.

However it should have two zeroes `0 0 .setfilladjust2`. Here's how it was explained back in 2001 when it was actually in the documentation:

> `<dx> <dy> .setfilladjust2 -`
> Sets graphics state parameters that cause all filled and stroked regions to be "fattened" by the given amount relative to an algorithm that only paints pixels whose centers fall within the region to be painted. dx and dy are numbers between 0 and 0.5, measured in device space. The only two values that are likely to be useful are 0, which gives a pure center-of-pixel rule, and 0.5, which gives Adobe's any-part-of-pixel rule. (0.5 is treated slightly specially in order to create half-open pixels per Adobe's specification.)

http://web.mit.edu/ghostscript/www/Language.htm

**Comment by tillkamppeter** [comment]
Done in commit c81914040b.
